Mathlib/Order/Filter/Pi.lean

```diff
@@ -0,0 +1,271 @@
+    choose x hxs hxt using this
+    exact hts (fun i _ => hxt i) (mem_univ_pi.2 hxs)
+  · simp only [inf_principal_eq_bot]
+    rintro ⟨i, hi⟩
+    filter_upwards [mem_pi_of_mem i hi]with x using mt fun h => h i trivial
+#align filter.pi_inf_principal_univ_pi_eq_bot Filter.pi_inf_principal_univ_pi_eq_bot
+
+@[simp]
+theorem pi_inf_principal_pi_eq_bot [∀ i, NeBot (f i)] {I : Set ι} :
+    pi f ⊓ 𝓟 (Set.pi I s) = ⊥ ↔ ∃ i ∈ I, f i ⊓ 𝓟 (s i) = ⊥ := by
+  rw [← univ_pi_piecewise I, pi_inf_principal_univ_pi_eq_bot]
+  refine' exists_congr fun i => _
+  by_cases hi : i ∈ I <;> simp [hi, NeBot.ne']
+#align filter.pi_inf_principal_pi_eq_bot Filter.pi_inf_principal_pi_eq_bot
+
+@[simp]
+theorem pi_inf_principal_univ_pi_neBot :
+    NeBot (pi f ⊓ 𝓟 (Set.pi univ s)) ↔ ∀ i, NeBot (f i ⊓ 𝓟 (s i)) := by simp [neBot_iff]
+#align filter.pi_inf_principal_univ_pi_ne_bot Filter.pi_inf_principal_univ_pi_neBot
+
+@[simp]
+theorem pi_inf_principal_pi_neBot [∀ i, NeBot (f i)] {I : Set ι} :
+    NeBot (pi f ⊓ 𝓟 (I.pi s)) ↔ ∀ i ∈ I, NeBot (f i ⊓ 𝓟 (s i)) := by simp [neBot_iff]
+#align filter.pi_inf_principal_pi_ne_bot Filter.pi_inf_principal_pi_neBot
+
+instance PiInfPrincipalPi.neBot [h : ∀ i, NeBot (f i ⊓ 𝓟 (s i))] {I : Set ι} :
+    NeBot (pi f ⊓ 𝓟 (I.pi s)) :=
+  (pi_inf_principal_univ_pi_neBot.2 ‹_›).mono <|
+    inf_le_inf_left _ <| principal_mono.2 fun x hx i _ => hx i trivial
+#align filter.pi_inf_principal_pi.ne_bot Filter.PiInfPrincipalPi.neBot
+
+@[simp]
+theorem pi_eq_bot : pi f = ⊥ ↔ ∃ i, f i = ⊥ := by
+  simpa using @pi_inf_principal_univ_pi_eq_bot ι α f fun _ => univ
+#align filter.pi_eq_bot Filter.pi_eq_bot
+
+@[simp]
+theorem pi_neBot : NeBot (pi f) ↔ ∀ i, NeBot (f i) := by simp [neBot_iff]
+#align filter.pi_ne_bot Filter.pi_neBot
+
+instance [∀ i, NeBot (f i)] : NeBot (pi f) :=
+  pi_neBot.2 ‹_›
+
+@[simp]
+theorem map_eval_pi (f : ∀ i, Filter (α i)) [∀ i, NeBot (f i)] (i : ι) :
+    map (eval i) (pi f) = f i := by
+  refine' le_antisymm (tendsto_eval_pi f i) fun s hs => _
+  rcases mem_pi.1 (mem_map.1 hs) with ⟨I, hIf, t, htf, hI⟩
+  rw [← image_subset_iff] at hI
+  refine' mem_of_superset (htf i) ((subset_eval_image_pi _ _).trans hI)
+  exact nonempty_of_mem (pi_mem_pi hIf fun i _ => htf i)
+#align filter.map_eval_pi Filter.map_eval_pi
+
+@[simp]
+theorem pi_le_pi [∀ i, NeBot (f₁ i)] : pi f₁ ≤ pi f₂ ↔ ∀ i, f₁ i ≤ f₂ i :=
+  ⟨fun h i => map_eval_pi f₁ i ▸ (tendsto_eval_pi _ _).mono_left h, pi_mono⟩
+#align filter.pi_le_pi Filter.pi_le_pi
+
+@[simp]
+theorem pi_inj [∀ i, NeBot (f₁ i)] : pi f₁ = pi f₂ ↔ f₁ = f₂ := by
+  refine' ⟨fun h => _, congr_arg pi⟩
+  have hle : f₁ ≤ f₂ := pi_le_pi.1 h.le
+  haveI : ∀ i, NeBot (f₂ i) := fun i => neBot_of_le (hle i)
+  exact hle.antisymm (pi_le_pi.1 h.ge)
+#align filter.pi_inj Filter.pi_inj
+
+end Pi
+
+/-! ### `n`-ary coproducts of filters -/
+
+section CoprodCat
+
+/-- Coproduct of filters. -/
+protected def Coprod (f : ∀ i, Filter (α i)) : Filter (∀ i, α i) :=
+  ⨆ i : ι, comap (eval i) (f i)
```

Comment:
I think this should become `coprod` + `subscript i`, just like `supr` became `sup_i` etc...

Reply:
Renamed.